### PR TITLE
[Bugfix] Fix output_ids containing input prompt tokens in engine.generate

### LINKED ISSUE #17240: [Bug] engine.generate output_ids contains tokens from the input prompt prefix
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When calling `engine.generate(input_ids=...)`, the returned `output_ids` contains tokens from the input prompt prefix, not just the generated tokens. However, the `text` field is correctly decoded. This causes a mismatch between `output_ids` and `text`. See "Reproduction" below for an example. 

## Root Cause
Note: I am not familiar with the slang codebase I used Opus 4.5 to try to find the bug, so take this with a grain of salt. 

In `sglang/srt/managers/detokenizer_manager.py`, line ~231:

```python
return BatchStrOut(
    ...
    output_ids=recv_obj.decode_ids,  # BUG: should not use decode_ids
    ...
)
```

The `decode_ids` is designed for incremental detokenization and intentionally includes prefix context (controlled by `surr_offset` in `init_incremental_detokenize()`). This is correct for computing the `text` output but incorrect for `output_ids`.

The actual generated token IDs (`req.output_ids`) are available in `scheduler_output_processor_mixin.py` but only used when `skip_tokenizer_init=True`.

### Suggested Fix

In `detokenizer_manager.py`, the `BatchStrOut` should use the actual output token IDs instead of `decode_ids`. The scheduler already computes this in `output_ids.append(req.output_ids[send_token_offset:])` - it just needs to be passed through to the final response.

## Workaround

Use `skip_tokenizer_init=True` when creating the engine:

```python
engine = sgl.Engine(model_path=MODEL_PATH, skip_tokenizer_init=True)
```

This returns correct `output_ids` but disables text decoding (requires manual decoding).

Another workaround is to manually remove the duplicated tokens:
```python
# The completion_tokens count is accurate
n_completion = output["meta_info"]["completion_tokens"]
correct_output_ids = output["output_ids"][-n_completion:]
```

### Reproduction

```python
import sglang as sgl
from transformers import AutoTokenizer
import nest_asyncio

nest_asyncio.apply()

# Use any model - the bug is in SGLang's output handling
MODEL_PATH = "Qwen/Qwen3-0.6B"

tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)
engine = sgl.Engine(model_path=MODEL_PATH)

# Create a simple prompt and tokenize it
prompt = "What is 2+2? Answer:"
input_ids = tokenizer.encode(prompt, add_special_tokens=True)

# Generate
output = engine.generate(
    input_ids=[input_ids],
    sampling_params={"max_new_tokens": 10, "temperature": 0}
)[0]

# Decode output_ids to compare with text
decoded_from_ids = tokenizer.decode(output["output_ids"])
returned_text = output["text"]

print(f"Returned text: {repr(returned_text)}")
print(f"Decoded from output_ids: {repr(decoded_from_ids)}")
print(f"Match: {decoded_from_ids == returned_text}")

# Show the mismatch
print(f"\nFirst 5 output_ids: {output['output_ids'][:5]}")
print(f"Decoded: {[tokenizer.decode([t]) for t in output['output_ids'][:5]]}")
```

The output:
```
Returned text: ' 4\n\nThe question is: What is the'
Decoded from output_ids: '+2? Answer: 4\n\nThe question is: What is the'
Match: False

First 5 output_ids: [10, 17, 30, 21806, 25]
Decoded: ['+', '2', '?', ' Answer', ':']
```

## Expected Behavior

`output_ids` should contain only the generated tokens, so:
```python
tokenizer.decode(output["output_ids"]) == output["text"]
```

## Actual Behavior

`output_ids` contains tokens from the **input prompt prefix** (used for incremental detokenization), causing:
- `output_ids` to have extra tokens at the beginning
- `tokenizer.decode(output["output_ids"]) != output["text"]`

The decoded output ids contain the tokens ` ['+', '2', '?', ' Answer', ':']` from the prompt. These tokens should not be returned by the engine, as they are part of the prompt and were not generated by the model. On the other hand, the output text is correct. 

This happens for every model, here is a sample conversation generated with 'gpt-oss-120b`:

Input:
```
<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2026-01-16

Reasoning: MEDIUM

# Valid channels: analysis, commentary, final. Channel must be included for every message.<|end|><|start|>developer<|message|># Instructions

You are a helpful math tutor.

<|end|><|start|>user<|message|>What is 2 + 2?<|end|><|start|>assistant<|channel|>final<|message|>2 + 2 equals 4.<|end|><|start|>user<|message|>What about 5 * 7?<|end|><|start|>assistant
``` 

Output:
```
7?<|end|><|start|>assistant<|channel|>analysis<|message|>The user asks "What about 5 * 7?" Simple multiplication: 35. Provide answer.<|end|><|start|>assistant<|channel|>final<|message|>\(5 \times 7 = 35\).<|return|>
``` 

Again we have tokens that belong to the prompt in the output. `7?<|end|><|start|>assistant` should not be there. 

### Environment

Python: 3.12.11 (main, Aug  6 2025, 22:58:09) [Clang 20.1.4 ]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H200
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.2, V12.2.140
CUDA Driver Version: 550.163.01
PyTorch: 2.8.0+cu128
sglang: 0.5.2
sgl_kernel: 0.3.9.post2
flashinfer_python: 0.3.1
triton: 3.4.0
transformers: 4.56.1
torchao: 0.9.0
numpy: 2.2.6
aiohttp: 3.13.3
fastapi: 0.128.0
hf_transfer: 0.1.9
huggingface_hub: 0.36.0
interegular: 0.3.3
modelscope: 1.33.0
orjson: 3.11.5
outlines: 0.1.11
packaging: 25.0
psutil: 7.2.1
pydantic: 2.12.5
python-multipart: 0.0.21
pyzmq: 27.1.0
uvicorn: 0.40.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.1.24
openai: 1.99.1
tiktoken: 0.12.0
anthropic: 0.71.0
litellm: Module Not Found
decord: 0.6.0
NVIDIA Topology: 
	[4mGPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NV18	0-47,96-143	0		N/A
GPU1	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NV18	0-47,96-143	0		N/A
GPU2	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NV18	0-47,96-143	0		N/A
GPU3	NV18	NV18	NV18	 X 	NV18	NV18	NV18	NV18	0-47,96-143	0		N/A
GPU4	NV18	NV18	NV18	NV18	 X 	NV18	NV18	NV18	48-95,144-191	1		N/A
GPU5	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NV18	48-95,144-191	1		N/A
GPU6	NV18	NV18	NV18	NV18	NV18	NV18	 X 	NV18	48-95,144-191	1		N/A
GPU7	NV18	NV18	NV18	NV18	NV18	NV18	NV18	 X 	48-95,144-191	1		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

Hypervisor vendor: KVM
ulimit soft: 1048576

### PR DESCRIPTION
**Pull request by karanb192**
## Summary

Fixes #17240

When using `engine.generate()` with `input_ids` parameter, the returned `output_ids` was incorrectly containing tokens from the input prompt prefix instead of only the generated tokens. The `text` field was correctly decoded, but `output_ids` contained extra tokens at the beginning that came from the input prompt.

### Root Cause
The `DecodeStatus` initialization in the detokenizer manager was assigning `decode_ids` as a reference rather than a copy. This could cause reference-related issues when the list is extended on subsequent calls during incremental streaming output.

### Changes
- Creates a copy of `decode_ids` when initializing `DecodeStatus` to prevent potential reference-related issues
- Adds a test to verify `output_ids` only contains generated tokens (not input prompt tokens)

## Test plan
- [ ] Run existing tests to ensure no regression
- [ ] Run new test `test_output_ids_no_prompt_prefix.py` to verify the fix
- [ ] Verify that `len(output_ids) == completion_tokens` after the fix

## Reproduction (from issue)
```python
from sglang import Engine
from transformers import AutoTokenizer

model_name = "Qwen/Qwen3-0.6B"
engine = Engine(model_path=model_name, random_seed=0)
tokenizer = AutoTokenizer.from_pretrained(model_name)

prompt = "What is 2+2? Answer:"
input_ids = tokenizer.encode(prompt, add_special_tokens=True)

output = engine.generate(
    input_ids=[input_ids],
    sampling_params={"max_new_tokens": 10, "temperature": 0}
)[0]

# Before fix: output_ids contains prompt tokens
# After fix: len(output_ids) == completion_tokens
assert len(output["output_ids"]) == output["meta_info"]["completion_tokens"]
```

### CODE REVIEW

**Comment by hnyls2002** [comment]
@karanb192 The fix here -- copying decode_ids instead of aliasing it in DecodeStatus -- is now on main at detokenizer_manager.py (decode_ids=list(recv_obj.decode_ids[i])), introduced by #25098. Closing as superseded; please reopen if you still observe output_ids containing prompt tokens on current main.
